internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go

```diff
@@ -137,20 +137,45 @@ func (s *Server) Register(obj *grpc.Server) {
 //
 //nolint:gocyclo
 func (s *Server) ApplyConfiguration(ctx context.Context, in *machine.ApplyConfigurationRequest) (*machine.ApplyConfigurationResponse, error) {
-	log.Printf("apply config request: immediate %v, on reboot %v", in.Immediate, in.OnReboot)
+	mode := in.Mode.String()
+
+	// TODO: remove in v0.16
+	switch {
+	case in.Immediate: //nolint:staticcheck
+		in.Mode = machine.ApplyConfigurationRequest_NO_REBOOT
+	case in.OnReboot: //nolint:staticcheck
+		in.Mode = machine.ApplyConfigurationRequest_STAGED
+	}
 
 	cfgProvider, err := s.Controller.Runtime().LoadAndValidateConfig(in.GetData())
 	if err != nil {
 		return nil, err
 	}
 
-	// --immediate
-	if in.Immediate {
+	autoApplyInfo := ""
+
+	//nolint:exhaustive
+	switch in.Mode {
+	// --mode=no-reboot
+	case machine.ApplyConfigurationRequest_NO_REBOOT:
 		if err = s.Controller.Runtime().CanApplyImmediate(cfgProvider); err != nil {
 			return nil, err
```

Comment:
we could wrap that with `gRPC` status code, so that error is easily distinguishable from the regular failure.